File: graphs/medium_breadth_first_search.py
# Breadth First Search

class Node:
  def __init__(self, name):
    self.children = []
    self.name = name

  def addChild(self, name):
    self.children.append(Node(name))
    return self

  def breadthFirstSearch(self, array):
    queue = [ self ]

    while queue:
      curr = queue.pop(0) # remove from idx 0

      array.append(curr.name)

      for child in curr.children:
        queue.append(child)

    return array

File: graphs/medium_breadth_first_search.js
// Breadth First Search
class Node {
  constructor(name) {
    this.name = name;
    this.children = [];
  }

  addChild(name) {
    this.children.push(new Node(name));
    return this;
  }

  breadthFirstSearch(array) {
    const queue = [ this ]

    while (queue.length) {
      const curr = queue.shift()

      array.push(curr.name)

      for (let child of curr.children) {
        queue.push(child)
      }
    }

    return array
  }
}
